src/Microsoft.AspNetCore.Razor/Compilation/TagHelpers/TagHelperDescriptorProvider.cs

```diff
@@ -95,39 +96,107 @@ public TagHelperDescriptorProvider(IEnumerable<TagHelperDescriptor> descriptors)
+                            var valueOffset = 0;
+                            if (afterEquals == '\'' || afterEquals == '"')
+                            {
+                                valueOffset = 1;
+                            }
+                            var valueStart = equalsIndex + 1 + valueOffset;
+                            var valueLength = requiredAttribute.Length - equalsIndex - valueOffset * 2 - 2;
+                            var requiredValue = requiredAttribute.Substring(valueStart, valueLength);
+                            var selectorCharacter = requiredAttribute[equalsIndex - 1];
+
+                            switch (selectorCharacter)
+                            {
+                                case '^': // Value starts with
+                                    if (!matchingAttributes.Any(attribute => attribute.Value.StartsWith(requiredValue, StringComparison.Ordinal)))
+                                    {
+                                        return false;
+                                    }
+                                    break;
+                                case '$': // Value ends with
+                                    if (!matchingAttributes.Any(attribute => attribute.Value.EndsWith(requiredValue, StringComparison.Ordinal)))
+                                    {
+                                        return false;
+                                    }
+                                    break;
+                                default: // Value equals
+                                    if (!matchingAttributes.Any(attribute => string.Equals(attribute.Value, requiredValue, StringComparison.Ordinal)))
+                                    {
+                                        return false;
+                                    }
+                                    break;
+                            }
+                        }
                     }
 
                     return true;
                 });
         }
 
+        public static string GetRequiredAttributeName(string originalAttributeName)
+        {
+            if (originalAttributeName.Length <= 2 || originalAttributeName[0] != '[')
+            {
+                return originalAttributeName;
+            }
+
+            string name;
+
+            // CSS Selector
+            var equalsIndex = originalAttributeName.IndexOf('=');
+            if (equalsIndex == -1)
+            {
+                // Can assume there's a matching ']' at the end because it has run through validation already.
+                name = originalAttributeName.Substring(1, originalAttributeName.Length - 2);
+
+                return name;
+            }
+
+            var nameOffset = (originalAttributeName[equalsIndex - 1] == '^' || originalAttributeName[equalsIndex - 1] == '$') ? -2 : -1;
+            name = originalAttributeName.Substring(1, equalsIndex + nameOffset);
```

Comment:
Requirement that `^=` and `$=` are kept together is correct according to the [CSS grammar](https://www.w3.org/TR/css3-selectors/#grammar). However whitespace is allowed before and after the identifier.